src/NLog/Internal/PropertiesDictionary.cs

```diff
@@ -0,0 +1,642 @@
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Internal
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Dictionary that combines the standard <see cref="LogEventInfo.Properties" /> with the
+    /// MessageTemplate-properties extracted from the <see cref="LogEventInfo.Message" />.
+    /// 
+    /// The <see cref="MessageProperties" /> are returned as the first items
+    /// in the collection, and in positional order.
+    /// </summary>
+    internal sealed class PropertiesDictionary : IDictionary<object, object>
+    {
+        struct PropertyValue
+        {
+            public readonly object Value;
+            public readonly bool MessageProperty;
+
+            public PropertyValue(object value, bool messageProperty)
+            {
+                Value = value;
+                MessageProperty = messageProperty;
+            }
+        }
+
+        private Dictionary<object, PropertyValue> _eventProperties;
+        private IList<MessageTemplateParameter> _messageProperties;
+        private DictionaryCollection _keyCollection;
+        private DictionaryCollection _valueCollection;
+        private IDictionary _eventContextAdapter;
+
+        /// <summary>
+        /// Injects the list of message-template-parameter into the IDictionary-interface
+        /// </summary>
+        /// <param name="parameterList">Message-template-parameters</param>
+        public PropertiesDictionary(IList<MessageTemplateParameter> parameterList = null)
+        {
+            MessageProperties = parameterList;
+        }
+
+        private bool IsEmpty { get { return (_eventProperties == null || _eventProperties.Count == 0) && (_messageProperties == null || _messageProperties.Count == 0); } }
+
+        public IDictionary EventContext { get { return _eventContextAdapter ?? (_eventContextAdapter = new DictionaryAdapter<object, object>(this)); } }
+
+        private Dictionary<object, PropertyValue> EventProperties
```

Comment:
These are the regular properties, isn;t?

Reply:
Not sure I understand the question.